src/librustc/hir/map/definitions.rs

```diff
@@ -575,6 +610,9 @@ impl Definitions {
         if node_id != ast::DUMMY_NODE_ID {
             debug!("create_def_with_parent: def_index_to_node[{:?} <-> {:?}", index, node_id);
             self.node_to_def_index.insert(node_id, index);
+            if let Some(hir_id) = self.node_to_hir_id.get(node_id) {
+                self.hir_to_def_index.insert(*hir_id, index);
+            }
```

Comment:
Does this not always produce a HirId?

Reply:
I was surprised too, but it doesn't - without this bit `stage1` won't build with an `index out of bounds: the len is 0 but the index is 108` error.

Reply:
Ah. The `Definitions` type is used before HIR lowering. So the HIR ids don't exist then. I'd suggest you remove the `hir_to_def_index` field and instead add utility methods to the HIR map which converts things to node ids. Use a separate commit for these changes. It makes it easier to review.